### PR TITLE
Make standard codecs extensible

### LINKED ISSUE #15389: StandardMessageCodec should be extensible
Working on issues with the Firestore plugin, specifically #13043, and the need to have support for additional types that are specific to the Firestore plugin, led to the creation of flutter/plugins#343 .

This solution works, but introduces code duplication and additional maintenance needs.

A more permanently correct solution would be to allow for StandardMessageCodec to be extended somehow. The particulars of how to do so are why I opened this for discussion.

### PR DESCRIPTION
**Pull request by mravn-google**
Make standard message/method codecs extensible. Awaits companion flutter/engine PR (https://github.com/flutter/engine/pull/4770).

Fixes #15389

### CODE REVIEW

File: packages/flutter/lib/src/services/message_codecs.dart [hunk 1]
```diff
@@ -309,53 +321,48 @@ class StandardMessageCodec implements MessageCodec<dynamic> {
...
       value.forEach((dynamic key, dynamic value) {
-        _writeValue(buffer, key);
-        _writeValue(buffer, value);
+        writeValue(buffer, key);
+        writeValue(buffer, value);
       });
     } else {
-      throw new ArgumentError.value(value);
+      buffer.putUint8(_kUnknown);
+      writeUnknown(buffer, value);
```
Comment (Hixie) [inline packages/flutter/lib/src/services/message_codecs.dart]: Another way to make this extensible would just be to have people override `writeValue` and call `super.writeValue` if they don't recognize the type.

Comment (Hixie) [reply packages/flutter/lib/src/services/message_codecs.dart]: we'd have to define that the format is that you first push a byte then push the data, and define a range of bytes that we reserve for our purposes.

File: packages/flutter/lib/src/services/message_codecs.dart [hunk 2]
```diff
@@ -309,53 +321,48 @@ class StandardMessageCodec implements MessageCodec<dynamic> {
...
-
-  static dynamic _readValue(ReadBuffer buffer) {
+  /// Reads a value from [buffer] as written by [writeValue].
+  ///
+  /// This method is intended for use by subclasses overriding
+  /// [readUnknown].
+  dynamic readValue(ReadBuffer buffer) {
     if (!buffer.hasRemaining)
       throw const FormatException('Message corrupted');
     dynamic result;
```
Comment (Hixie) [inline packages/flutter/lib/src/services/message_codecs.dart]: similarly here we could read out the first byte, then call readValue with that byte and the buffer, and they would override readValue and call super.readValue if they don't recognise the type

**Comment by Hixie** [comment]
My main concern with the readUnknown/writeUnknown thing is that it makes extensions into second class citizens rather than treating them the same way we treat our own code.

**Comment by mravn-google** [comment]
> My main concern with the readUnknown/writeUnknown thing is that it makes extensions into second class citizens rather than treating them the same way we treat our own code.

Roger that. My main concern was to protect the range of type discriminator bytes used by the base class. Documenting our way out of that makes for a more straightforward extension API. Done.

File: packages/flutter/lib/src/services/message_codecs.dart [hunk 3]
```diff
@@ -271,26 +274,23 @@ class StandardMessageCodec implements MessageCodec<dynamic> {
...
+  /// byte, then the value itself.
+  ///
+  /// Type discriminators 0 through 127 inclusive are reserved for use by the
+  /// base class.
+  ///
+  /// The codec can be extended by overriding this method, calling super
+  /// for values that the extension does not handle. Type discriminators
+  /// used by extensions must be greater than or equal to 128 in order to avoid
+  /// clashes with any later extensions to the base class.
+  void writeValue(WriteBuffer buffer, dynamic value) {
```
Comment (Hixie) [inline packages/flutter/lib/src/services/message_codecs.dart]: maybe mention that it's ok to call this recursively, e.g. to implement writing a list of values.

File: packages/flutter/lib/src/services/message_codecs.dart [hunk 4]
```diff
@@ -201,6 +201,9 @@ class JSONMethodCodec implements MethodCodec {
 ///    `FlutterStandardTypedData`
 ///  * [List]\: `NSArray`
 ///  * [Map]\: `NSDictionary`
+///
+/// The codec is extensible by subclasses overriding [writeValue] and
+/// [readValue].
```
Comment (Hixie) [inline packages/flutter/lib/src/services/message_codecs.dart]: readValueOfType?

File: packages/flutter/lib/src/services/message_codecs.dart [hunk 5]
```diff
@@ -379,51 +382,83 @@ class StandardMessageCodec implements MessageCodec<dynamic> {
...
+  /// This method is intended for use by subclasses overriding
+  /// [readValueOfType].
+  int readSize(ReadBuffer buffer) {
+    final int value = buffer.getUint8();
+    if (value < 254)
+      return value;
+    else if (value == 254)
+      return buffer.getUint16();
+    else
+      return buffer.getUint32();
```
Comment (Hixie) [inline packages/flutter/lib/src/services/message_codecs.dart]: this might be more readable as a switch

**Comment by Hixie** [comment]
LGTM looks great.

**Pull request merged**